src/azul/plugins/__init__.py

```diff
@@ -191,16 +195,31 @@ def sources(self) -> AbstractSet[SOURCE_SPEC]:
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def list_sources(self,
+                     authentication: Optional[Authentication]
+                     ) -> Iterable[SOURCE_REF]:
+        """
+        List all sources in the underlying repository. Implementations should
+        raise RequirementError if the provided authentication is insufficient
+        to access the repository.
+        """
+        raise NotImplementedError
+
+    def source_classes(self) -> Tuple[Type[SOURCE_SPEC], Type[SOURCE_REF]]:
```

Comment:
This is a rather complicated concern that ideally I wouldn't want to bother clients with. It's ok to have this magic as an implementation detail of a method, but exposing it externally complicates the client interface and is brittle.